src/system/process.py

```diff
@@ -0,0 +1,51 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import logging
+import os
+import subprocess
+
+import psutil  # type: ignore
+
+
+def terminate(process, work_dir, stdout, stderr):
+    parent = psutil.Process(process.pid)
+    logging.debug("Checking for child processes")
+    child_processes = parent.children(recursive=True)
+    for child in child_processes:
+        logging.debug(f"Found child process with pid {child.pid}")
+        if child.pid != process.pid:
+            logging.debug(f"Sending SIGKILL to {child.pid} ")
+            child.kill()
+    logging.info(f"Sending SIGTERM to PID {process.pid}")
+    process.terminate()
+    try:
+        logging.info("Waiting for process to terminate")
+        process.wait(10)
+    except subprocess.TimeoutExpired:
+        logging.info("Process did not terminate after 10 seconds. Sending SIGKILL")
+        process.kill()
+        try:
+            logging.info("Waiting for process to terminate")
+            process.wait(10)
+        except subprocess.TimeoutExpired:
+            logging.info("Process failed to terminate even after SIGKILL")
+            raise
+    finally:
+        logging.info(f"Process terminated with exit code {process.returncode}")
+        if stdout:
```

Comment:
This logic is specific to test cluster code and should be either moved out to the calling code or refactored to be generic. I think you want to wrap the process in a `Process` class that can deal with its starting and stopping, stderr and stdout.

Reply:
Addressed